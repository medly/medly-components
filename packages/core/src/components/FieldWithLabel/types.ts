import { Theme } from '@medly-components/theme';
import { HTMLProps } from '@medly-components/utils';
import { StyledComponent } from 'styled-components';
import Label from '../Label';
import { LabelPositions } from '../Label/types';
import Text from '../Text';

export interface FieldWithLabelProps extends HTMLProps<HTMLDivElement> {
    /** Label Position */
    labelPosition?: LabelPositions;
    /** Takes full width of the parent component */
    fullWidth?: boolean;
    /** To be used when field should take max width */
    fieldWithMaxContent?: boolean;
    /** Min width of the component */
    minWidth?: number;
}

export interface FieldWithLabelStyledProps extends FieldWithLabelProps {
    isLabelPresent?: boolean;
}

export interface FieldStyledProps {
    isIndented?: boolean;
}

export interface StaticProps {
    Field: StyledComponent<'div', Theme, FieldStyledProps>;
    Label: StyledComponent<typeof Label, Theme>;
    Description: StyledComponent<typeof Text, Theme>;
}
