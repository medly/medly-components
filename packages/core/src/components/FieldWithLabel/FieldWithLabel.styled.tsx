import { defaultTheme } from '@medly-components/theme';
import { css, styled } from '@medly-components/utils';
import { AnyStyledComponent } from 'styled-components';
import Label from '../Label';
import Text from '../Text';
import { FieldStyledProps, FieldWithLabelStyledProps } from './types';

export const FieldStyled = styled('div')<FieldStyledProps>`
    grid-area: field;
    flex-wrap: wrap;
    margin-left: ${({ isIndented }) => (isIndented ? '17px' : '0px')};
`;

export const LabelStyled = styled(Label)`
    grid-area: label;
    align-self: center;
    user-select: none;
`;
LabelStyled.displayName = 'Label';

export const DescriptionStyled = styled(Text)`
    grid-area: description;
    margin: 5px 0 0 2px;
    color: ${({ theme, textColor }) => (textColor ? textColor : theme.input.descriptionColor)};
`;

export const FieldWithLabelStyled: AnyStyledComponent = styled('div')<FieldWithLabelStyledProps>`
    margin: ${({ theme, fullWidth }) =>
        fullWidth ? `${theme.spacing.S2} 0` : `${theme.spacing.S2} ${theme.spacing.S2} ${theme.spacing.S2} 0`};
    display: ${({ fullWidth }) => (fullWidth ? 'grid' : 'inline-grid')};
    align-items: center;
    ${({ minWidth }) =>
        minWidth &&
        css`
            min-width: ${minWidth}px;
        `}
    ${FieldStyled} {
        display: ${({ fullWidth }) => (fullWidth ? 'flex' : 'inline-flex')};
        align-items: ${({ labelPosition }) => (labelPosition === 'top' || labelPosition === 'bottom' ? 'flex-start' : 'center')};
        flex-direction: ${({ labelPosition }) => {
            switch (labelPosition) {
                case 'bottom':
                    return `column-reverse`;
                case 'left':
                    return `row`;
                case 'right':
                    return `row-reverse`;
                case 'top':
                default:
                    return 'column';
            }
        }};
    }
    ${({ labelPosition, fieldWithMaxContent }) => {
        switch (labelPosition) {
            case 'top':
                return css`
                    grid-template-areas:
                        'label'
                        'field'
                        'description';
                `;
            case 'bottom':
                return css`
                    grid-template-areas:
                        'field'
                        'description'
                        'label';
                `;
            case 'left':
                return css`
                    grid-template-columns: ${fieldWithMaxContent
                        ? 'minmax(max-content, 25%)  max-content'
                        : 'minmax(max-content, 25%) minmax(auto, 80%)'};
                    grid-template-areas:
                        'label field'
                        'random description';
                `;
            case 'right':
            default:
                return css`
                    grid-template-columns: ${fieldWithMaxContent
                        ? 'max-content minmax(max-content, 25%)'
                        : 'minmax(max-content, 80%) minmax(max-content, 25%)'};
                    grid-template-areas:
                        'field label'
                        'description random';
                `;
        }
    }};
    ${({ isLabelPresent }) =>
        !isLabelPresent &&
        css`
            grid-template-columns: auto;
            grid-template-areas:
                'field'
                'description';
        `}
`;
FieldWithLabelStyled.defaultProps = {
    theme: defaultTheme
};
