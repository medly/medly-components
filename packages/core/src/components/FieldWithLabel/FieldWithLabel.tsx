import { WithStyle } from '@medly-components/utils';
import { Children, useMemo, memo, forwardRef } from 'react';
import { DescriptionStyled, FieldStyled, FieldWithLabelStyled, LabelStyled } from './FieldWithLabel.styled';
import { FieldWithLabelProps, StaticProps } from './types';
import type { FC } from 'react';

const Component: FC<FieldWithLabelProps> = memo(
    forwardRef((props, ref) => {
        const isLabelPresent = useMemo(
            () => !!Children.toArray(props.children).find((child: any) => child && child.type && child.type.displayName === 'Label'),
            [props.children]
        );

        return <FieldWithLabelStyled ref={ref} {...props} isLabelPresent={isLabelPresent} />;
    })
);

Component.displayName = 'FieldWithLabel';
Component.defaultProps = { labelPosition: 'left', fullWidth: false, fieldWithMaxContent: false };
export const FieldWithLabel: FC<FieldWithLabelProps> & WithStyle & StaticProps = Object.assign(Component, {
    Style: FieldWithLabelStyled,
    Description: DescriptionStyled,
    Label: LabelStyled,
    Field: FieldStyled
});
