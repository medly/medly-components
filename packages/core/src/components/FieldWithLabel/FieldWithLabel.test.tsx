import { render } from '@test-utils';
import { FieldWithLabel } from './FieldWithLabel';

describe('FieldWithLabel component', () => {
    test.each`
        labelPosition | fullWidth | fieldWithMaxContent
        ${'left'}     | ${true}   | ${true}
        ${'right'}    | ${true}   | ${false}
        ${'top'}      | ${false}  | ${true}
        ${'bottom'}   | ${false}  | ${false}
    `(
        'should render properly with props labelPosition: $labelPosition, fullWidth: $fullWidth, fieldWithMaxContent: $fieldWithMaxContent',
        props => {
            const { container } = render(
                <FieldWithLabel {...props} minWidth={300}>
                    <FieldWithLabel.Label>Dummy Label</FieldWithLabel.Label>
                    <FieldWithLabel.Field>Dummy Field</FieldWithLabel.Field>
                    <FieldWithLabel.Description>Dummy Description</FieldWithLabel.Description>
                </FieldWithLabel>
            );
            expect(container).toMatchSnapshot();
        }
    );

    it('should render properly when lable is not present', () => {
        const { container } = render(
            <FieldWithLabel>
                <FieldWithLabel.Field isIndented>Dummy Field</FieldWithLabel.Field>
                <FieldWithLabel.Description>Dummy Description</FieldWithLabel.Description>
            </FieldWithLabel>
        );
        expect(container).toMatchSnapshot();
    });
});
