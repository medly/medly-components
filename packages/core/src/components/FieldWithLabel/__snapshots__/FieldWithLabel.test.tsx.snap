// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldWithLabel component should render properly when lable is not present 1`] = `
.c3 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c2 {
  grid-area: field;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-left: 17px;
}

.c4 {
  grid-area: description;
  margin: 5px 0 0 2px;
  color: #012040;
}

.c0 {
  margin: 0.8rem 0.8rem 0.8rem 0;
  display: inline-grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  grid-template-columns: minmax(max-content,25%) minmax(auto,80%);
  grid-template-areas: 'label field' 'random description';
  grid-template-columns: auto;
  grid-template-areas: 'field' 'description';
}

.c0 .c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1 c2"
    >
      Dummy Field
    </div>
    <span
      class="c3 c4"
    >
      Dummy Description
    </span>
  </div>
</div>
`;

exports[`FieldWithLabel component should render properly with props labelPosition: bottom, fullWidth: false, fieldWithMaxContent: false 1`] = `
.c1 {
  color: #13181D;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  margin: 0px 0.8rem 0px 0px;
  cursor: default;
}

.c5 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c4 {
  grid-area: field;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-left: 0px;
}

.c2 {
  grid-area: label;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c6 {
  grid-area: description;
  margin: 5px 0 0 2px;
  color: #012040;
}

.c0 {
  margin: 0.8rem 0.8rem 0.8rem 0;
  display: inline-grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 300px;
  grid-template-areas: 'field' 'description' 'label';
}

.c0 .c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex-direction: column-reverse;
  -ms-flex-direction: column-reverse;
  flex-direction: column-reverse;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1 c2"
    >
      Dummy Label
    </label>
    <div
      class="c3 c4"
    >
      Dummy Field
    </div>
    <span
      class="c5 c6"
    >
      Dummy Description
    </span>
  </div>
</div>
`;

exports[`FieldWithLabel component should render properly with props labelPosition: left, fullWidth: true, fieldWithMaxContent: true 1`] = `
.c1 {
  color: #13181D;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  margin: 0px 0.8rem 0px 0px;
  cursor: default;
}

.c5 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c4 {
  grid-area: field;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-left: 0px;
}

.c2 {
  grid-area: label;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c6 {
  grid-area: description;
  margin: 5px 0 0 2px;
  color: #012040;
}

.c0 {
  margin: 0.8rem 0;
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 300px;
  grid-template-columns: minmax(max-content,25%) max-content;
  grid-template-areas: 'label field' 'random description';
}

.c0 .c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1 c2"
    >
      Dummy Label
    </label>
    <div
      class="c3 c4"
    >
      Dummy Field
    </div>
    <span
      class="c5 c6"
    >
      Dummy Description
    </span>
  </div>
</div>
`;

exports[`FieldWithLabel component should render properly with props labelPosition: right, fullWidth: true, fieldWithMaxContent: false 1`] = `
.c1 {
  color: #13181D;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  margin: 0px 0.8rem 0px 0px;
  cursor: default;
}

.c5 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c4 {
  grid-area: field;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-left: 0px;
}

.c2 {
  grid-area: label;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c6 {
  grid-area: description;
  margin: 5px 0 0 2px;
  color: #012040;
}

.c0 {
  margin: 0.8rem 0;
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 300px;
  grid-template-columns: minmax(max-content,80%) minmax(max-content,25%);
  grid-template-areas: 'field label' 'description random';
}

.c0 .c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1 c2"
    >
      Dummy Label
    </label>
    <div
      class="c3 c4"
    >
      Dummy Field
    </div>
    <span
      class="c5 c6"
    >
      Dummy Description
    </span>
  </div>
</div>
`;

exports[`FieldWithLabel component should render properly with props labelPosition: top, fullWidth: false, fieldWithMaxContent: true 1`] = `
.c1 {
  color: #13181D;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  margin: 0px 0.8rem 0px 0px;
  cursor: default;
}

.c5 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c4 {
  grid-area: field;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-left: 0px;
}

.c2 {
  grid-area: label;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c6 {
  grid-area: description;
  margin: 5px 0 0 2px;
  color: #012040;
}

.c0 {
  margin: 0.8rem 0.8rem 0.8rem 0;
  display: inline-grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 300px;
  grid-template-areas: 'label' 'field' 'description';
}

.c0 .c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1 c2"
    >
      Dummy Label
    </label>
    <div
      class="c3 c4"
    >
      Dummy Field
    </div>
    <span
      class="c5 c6"
    >
      Dummy Description
    </span>
  </div>
</div>
`;
